src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+    /// bool)"/> except that it will still fork even if <paramref name="newSolutionState"/> is unchanged from
+    /// <see cref="Solution"/>.
+    /// </summary>
+    private SolutionCompilationState ForceForkProject(
+        SolutionState newSolutionState,
+        ProjectState newProjectState,
+        CompilationAndGeneratorDriverTranslationAction? translate,
+        bool forkTracker)
+    {
+        var projectId = newProjectState.Id;
+
+        var newDependencyGraph = newSolutionState.GetProjectDependencyGraph();
+        var newTrackerMap = CreateCompilationTrackerMap(projectId, newDependencyGraph);
+
+        // If we have a tracker for this project, then fork it as well (along with the
+        // translation action and store it in the tracker map.
+        if (newTrackerMap.TryGetValue(projectId, out var tracker))
+        {
+            newTrackerMap = newTrackerMap.Remove(projectId);
+
+            if (forkTracker)
+            {
+                newTrackerMap = newTrackerMap.Add(projectId, tracker.Fork(newProjectState, translate));
+            }
+        }
+
+        return this.Branch(
+            newSolutionState,
+            projectIdToTrackerMap: newTrackerMap);
+    }
+
+    private ImmutableDictionary<ProjectId, ICompilationTracker> CreateCompilationTrackerMap(ProjectId changedProjectId, ProjectDependencyGraph dependencyGraph)
+    {
+        if (_projectIdToTrackerMap.Count == 0)
+            return _projectIdToTrackerMap;
+
+        using var _ = ArrayBuilder<KeyValuePair<ProjectId, ICompilationTracker>>.GetInstance(_projectIdToTrackerMap.Count, out var newTrackerInfo);
+        var allReused = true;
+        foreach (var (id, tracker) in _projectIdToTrackerMap)
+        {
+            var localTracker = tracker;
+            if (!CanReuse(id))
+            {
+                localTracker = tracker.Fork(tracker.ProjectState, translate: null);
+                allReused = false;
+            }
+
+            newTrackerInfo.Add(new KeyValuePair<ProjectId, ICompilationTracker>(id, localTracker));
+        }
+
+        if (allReused)
+            return _projectIdToTrackerMap;
+
+        return ImmutableDictionary.CreateRange(newTrackerInfo);
+
+        // Returns true if 'tracker' can be reused for project 'id'
+        bool CanReuse(ProjectId id)
+        {
+            if (id == changedProjectId)
+            {
+                return true;
+            }
+
+            return !dependencyGraph.DoesProjectTransitivelyDependOnProject(id, changedProjectId);
+        }
+    }
+
+    /// <inheritdoc cref="SolutionState.AddProject(ProjectInfo)"/>
+    public SolutionCompilationState AddProject(ProjectInfo projectInfo)
+    {
+        var newSolutionState = this.Solution.AddProject(projectInfo);
+        var newTrackerMap = CreateCompilationTrackerMap(projectInfo.Id, newSolutionState.GetProjectDependencyGraph());
```

Comment:
an example of a line of code that used to be in SOlutionState.AddProejct, but moved here since this is part of the state of SolutionCompilationState